src/Workspaces/Core/Portable/Workspace/Host/SourceFiles/IDynamicFileInfoProvider.cs

```diff
@@ -0,0 +1,33 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.Host
+{
+    internal interface IDynamicFileInfoProvider
+    {
+        /// <summary>
+        /// return <see cref="DynamicFileInfo"/> for the context given
+        /// </summary>
+        /// <param name="projectId"><see cref="ProjectId"/> this file belongs to</param>
+        /// <param name="projectFilePath">full path to project file (ex, csproj)</param>
+        /// <param name="filePath">full path to non source file (ex, cshtml)</param>
+        /// <returns>null if this provider can't handle the given file</returns>
+        Task<DynamicFileInfo> GetDynamicFileInfoAsync(ProjectId projectId, string projectFilePath, string filePath, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// let provider know certain file has been removed
+        /// </summary>
+        /// <param name="projectId"><see cref="ProjectId"/> this file belongs to</param>
+        /// <param name="projectFilePath">full path to project file (ex, csproj)</param>
+        /// <param name="filePath">full path to non source file (ex, cshtml)</param>
+        Task RemoveDynamicFileInfoAsync(ProjectId projectId, string projectFilePath, string filePath, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// indicate content of a file has updated. the event argument "string" should be same as "filepath" given to <see cref="GetDynamicFileInfoAsync(ProjectId, string, string, CancellationToken)"/>
+        /// </summary>
+        event EventHandler<string> Updated;
```

Comment:
👍 Doing anything with strings/paths seems enormously fragile.  Can we not have a DynamicDocumentId for this sort of thing, similar to our DocumentId?

Reply:
we could ask them to give us new DynamicFileInfo. not in this PR, but sure. @rynowak

Reply:
i'm a little confused why we wouldn't just do these things in this PR.  This is the code where we're adding this support.  So why not just impl things with a non-fragile design, and then when this flows to our partner teams they can use appropriately.  They'd have to do this anyways just to use the above API.  So it seems fine to just have them use DynamicFileInfo instead of a string.